![banner](../banner.jpg)

***[Workshop 2](about) - Training course in data analysis for genomic surveillance of African malaria vectors***

---

# Module 1 - Introduction to pandas DataFrames

**Theme: Tools & Technology**

In this module we're going to take a closer look at a Python package called [pandas](https://pandas.pydata.org/), which is an essential part of our data analysis toolkit. We use pandas whenever we need to store, manipulate or analyse tables of data, also known as DataFrames. In this module we'll focus on using pandas to analyse genomic data, but it can be used to analyse any kind of data, so it's a very useful tool to have at your fingertips.      

## Learning objectives 

After this module you will be able to:

* Explain what a DataFrame is
* Access columns and rows from a DataFrame
* Use indexes and queries
* Write a DataFrame to a file

Please note that this module will cover only a small part of the features that pandas provides - see the links at then end of this notebook if you'd like to go further.

## Lecture

### English

In [ ]:
%%html
<iframe width="560" height="315" src="https://www.youtube.com/embed/iIYftkvS0eM" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

### Français 

In [ ]:
%%html
<iframe width="560" height="315" src="https://www.youtube.com/embed/6ycnU50kzZI" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>

## Setup

Let's install the `malariagen_data` package so we have easy access to some genomic data to demonstrate with. Pandas is already installed on colab.

In [1]:
!pip install -q malariagen_data

     |████████████████████████████████| 100 kB 2.6 MB/s 
     |████████████████████████████████| 3.6 MB 41.2 MB/s 
     |████████████████████████████████| 301 kB 46.1 MB/s 
     |████████████████████████████████| 3.0 MB 46.5 MB/s 
     |████████████████████████████████| 9.9 MB 37.9 MB/s 
     |████████████████████████████████| 3.7 MB 39.8 MB/s 
     |████████████████████████████████| 5.7 MB 42.8 MB/s 
     |████████████████████████████████| 185 kB 46.7 MB/s 
     |████████████████████████████████| 55 kB 2.3 MB/s 
     |████████████████████████████████| 1.6 MB 54.6 MB/s 
     |████████████████████████████████| 6.6 MB 22.0 MB/s 


Now import the `malariagen_data` and `pandas` packages. Note that by convention we use `pd` as an abbreviation for `pandas`.

In [2]:
import malariagen_data
import pandas as pd

Set up access to the MalariaGEN Ag3.0 data resource.

In [3]:
ag3 = malariagen_data.Ag3()
ag3

<MalariaGEN Ag3 API client>
Storage URL             : gs://vo_agam_release/
Data releases available : 3.0
Results cache           : None
Cohorts analysis        : 20220608
Species analysis        : aim_20220528
Site filters analysis   : dt_20200416
Software version        : malariagen_data 7.0.4
Client location         : Iowa, US (colab)
---
Please note that data are subject to terms of use,
for more information see https://www.malariagen.net/data
or contact data@malariagen.net. For API documentation see 
https://malariagen.github.io/vector-data/ag3/api.html

## pandas DataFrames

We use the pandas `DataFrame` class whenever we want to store and analyse data that can be structured as a table with rows and columns. Let's take a look at some examples of DataFrames that we met in the [first workshop](https://anopheles-genomic-surveillance.github.io/workshop-1/about.html) of this training course.

### Example: sample metadata

In [workshop 1, module 2](https://anopheles-genomic-surveillance.github.io/workshop-1/module-2-sample-metadata.html) we encountered the [sample_metadata()](https://malariagen.github.io/vector-data/ag3/api.html#sample-metadata) function. This function returns a DataFrame of sample metadata. Let's call this function and assign the result to the variable `df_samples`.

In [4]:
df_samples = ag3.sample_metadata(sample_sets="3.0")

Load sample metadata:   0%|          | 0/28 [00:00<?, ?it/s]

Check what type of object we have received.

In [5]:
type(df_samples)

pandas.core.frame.DataFrame

Inspect the data in the DataFrame by running a code cell containing just the name of the variable.

In [6]:
df_samples

sample_id partner_sample_id      contributor    country location  year  \
0     AR0047-C            LUA047       Joao Pinto     Angola   Luanda  2009   
1     AR0049-C            LUA049       Joao Pinto     Angola   Luanda  2009   
2     AR0051-C            LUA051       Joao Pinto     Angola   Luanda  2009   
3     AR0061-C            LUA061       Joao Pinto     Angola   Luanda  2009   
4     AR0078-C            LUA078       Joao Pinto     Angola   Luanda  2009   
...        ...               ...              ...        ...      ...   ...   
3076  AD0494-C         80-2-o-16  Martin Donnelly  Lab Cross     LSTM    -1   
3077  AD0495-C         80-2-o-17  Martin Donnelly  Lab Cross     LSTM    -1   
3078  AD0496-C         80-2-o-18  Martin Donnelly  Lab Cross     LSTM    -1   
3079  AD0497-C         80-2-o-19  Martin Donnelly  Lab Cross     LSTM    -1   
3080  AD0498-C         80-2-o-20  Martin Donnelly  Lab Cross     LSTM    -1   

      month  latitude  longitude sex_call  ...                    aim_species  \
0         4    -8.884     13.302        F  ...                       coluzzii   
1         4    -8.884     13.302        F  ...                       coluzzii   
2         4    -8.884     13.302        F  ...                       coluzzii   
3         4    -8.884     13.302        F  ...                       coluzzii   
4         4    -8.884     13.302        F  ...                       coluzzii   
...     ...       ...        ...      ...  ...                            ...   
3076     -1    53.409     -2.969        F  ...  intermediate_gambiae_coluzzii   
3077     -1    53.409     -2.969        M  ...  intermediate_gambiae_coluzzii   
3078     -1    53.409     -2.969        M  ...  intermediate_gambiae_coluzzii   
3079     -1    53.409     -2.969        F  ...  intermediate_gambiae_coluzzii   
3080     -1    53.409     -2.969        M  ...  intermediate_gambiae_coluzzii   

     country_iso  admin1_name  admin1_iso  admin2_name  \
0            AGO       Luanda      AO-LUA       Luanda   
1            AGO       Luanda      AO-LUA       Luanda   
2            AGO       Luanda      AO-LUA       Luanda   
3            AGO       Luanda      AO-LUA       Luanda   
4            AGO       Luanda      AO-LUA       Luanda   
...          ...          ...         ...          ...   
3076         NaN          NaN         NaN          NaN   
3077         NaN          NaN         NaN          NaN   
3078         NaN          NaN         NaN          NaN   
3079         NaN          NaN         NaN          NaN   
3080         NaN          NaN         NaN          NaN   

                              taxon cohort_admin1_year  cohort_admin1_month  \
0                          coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
1                          coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
2                          coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
3                          coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
4                          coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
...                             ...                ...                  ...   
3076  intermediate_gambiae_coluzzii                NaN                  NaN   
3077  intermediate_gambiae_coluzzii                NaN                  NaN   
3078  intermediate_gambiae_coluzzii                NaN                  NaN   
3079  intermediate_gambiae_coluzzii                NaN                  NaN   
3080  intermediate_gambiae_coluzzii                NaN                  NaN   

           cohort_admin2_year         cohort_admin2_month  
0     AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
1     AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
2     AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
3     AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
4     AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
...                       ...                         ...  
3076             

There are 3,081 rows and 26 columns in this DataFrame. Each row provides data about a single mosquito DNA sample that has been sequenced by MalariaGEN. The columns provide data about those mosquitoes, such as the country and year in which the mosquito was originally collected.

Note that the representation of the DataFrame above only shows some of the data - there are too many rows and columns to show all at once. The ellipsis character ("...") shows where some data have been hidden.

### Example: genome annotations

In [workshop 1, module 3](https://anopheles-genomic-surveillance.github.io/workshop-1/module-3-genome.html) we encountered the [geneset()](https://malariagen.github.io/vector-data/ag3/api.html#geneset) function. This function returns a DataFrame of genome annotations. Let's call the function and assign the result to the variable `df_geneset`.

In [7]:
df_geneset = ag3.geneset()
df_geneset

contig      source             type   start       end  score  \
0               2L  VectorBase       chromosome       1  49364325    NaN   
1               2L  VectorBase             gene  157348    186936    NaN   
2               2L  VectorBase             mRNA  157348    181305    NaN   
3               2L  VectorBase  three_prime_UTR  157348    157495    NaN   
4               2L  VectorBase             exon  157348    157623    NaN   
...            ...         ...              ...     ...       ...    ...   
196140  Y_unplaced  VectorBase   five_prime_UTR   47932     48111    NaN   
196141  Y_unplaced  VectorBase             exon   47932     48138    NaN   
196142  Y_unplaced  VectorBase              CDS   48112     48138    NaN   
196143  Y_unplaced  VectorBase             exon   48301     48385    NaN   
196144  Y_unplaced  VectorBase              CDS   48301     48385    NaN   

       strand  phase             ID         Parent              Name  \
0         NaN    NaN             2L            NaN               NaN   
1           -    NaN     AGAP004677            NaN               NaN   
2           -    NaN  AGAP004677-RA     AGAP004677               NaN   
3           -    NaN            NaN  AGAP004677-RA               NaN   
4           -    NaN            NaN  AGAP004677-RA  AGAP004677-RB-E4   
...       ...    ...            ...            ...               ...   
196140      +    NaN            NaN  AGAP029375-RA               NaN   
196141      +    NaN            NaN  AGAP029375-RA  AGAP029375-RA-E2   
196142      +    0.0  AGAP029375-PA  AGAP029375-RA               NaN   
196143      +    NaN            NaN  AGAP029375-RA  AGAP029375-RA-E3   
196144      +    0.0  AGAP029375-PA  AGAP029375-RA               NaN   

                                              description  
0                                                     NaN  
1       methylenetetrahydrofolate dehydrogenase(NAD ) ...  
2                                                     NaN  
3                                                     NaN  
4                                                     NaN  
...                                                   ...  
196140                                                NaN  
196141                                                NaN  
196142                                                NaN  
196143                                                NaN  
196144                                                NaN  

[196145 rows x 12 columns]

There are 196,145 rows and 12 columns in this DataFrame. Each row provides data about an annotated feature on the *Anopheles gambiae* AgamP4 reference genome, such as a gene. The columns provide data about those features, such as the contig (chromosome arm) on which the feature occurs, and the start and end coordinates. 

### Example: SNP allele frequencies

In [workshop 1, module 4](https://anopheles-genomic-surveillance.github.io/workshop-1/module-4-vgsc-snps.html), we encountered the [snp_allele_frequencies()](https://malariagen.github.io/vector-data/ag3/api.html#snp-allele-frequencies) function. This function returns a DataFrame of single nucleotide polymorphisms (SNPs) together with their allele frequencies in different mosquito cohorts. Let's call this function and assign the result to the variable `snp_freqs_df`.

In [8]:
snp_freqs_df = ag3.snp_allele_frequencies(
    transcript="AGAP004707-RD", 
    cohorts="admin1_year", 
    sample_sets=["AG1000G-BF-A", "AG1000G-BF-B", "AG1000G-BF-C"], 
)
snp_freqs_df

Load sample metadata:   0%|          | 0/3 [00:00<?, ?it/s]

Load SNP genotypes:   0%|          | 0/32 [00:00<?, ?it/s]

Compute allele frequencies:   0%|          | 0/6 [00:00<?, ?it/s]

Compute SNP effects:   0%|          | 0/9659 [00:00<?, ?it/s]

pass_gamb_colu_arab  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                      True   
       2358328  T          C          NaN                       True   
       2358441  A          T          NaN                       True   
       2358468  A          C          NaN                       True   
       2358581  T          A          NaN                       True   
...                                                              ...   
       2431424  G          T          G2055V                    True   
       2431518  C          A          A2086A                    True   
       2431527  G          C          V2089V                    True   
       2431569  C          A          V2103V                    True   
       2431613  T          C          V2118A                    True   

                                                 pass_gamb_colu  pass_arab  \
contig position ref_allele alt_allele aa_change                              
2L     2358252  C          T          A32V                 True       True   
       2358328  T          C          NaN                  True       True   
       2358441  A          T          NaN                  True       True   
       2358468  A          C          NaN                  True       True   
       2358581  T          A          NaN                  True       True   
...                                                         ...        ...   
       2431424  G          T          G2055V               True       True   
       2431518  C          A          A2086A               True       True   
       2431527  G          C          V2089V               True       True   
       2431569  C          A          V2103V               True       True   
       2431613  T          C          V2118A               True       True   

                                                 frq_BF-07_gamb_2004  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                  0.000000   
       2358328  T          C          NaN                   0.000000   
       2358441  A          T          NaN                   0.038462   
       2358468  A          C          NaN                   0.269231   
       2358581  T          A          NaN                   0.192308   
...                                                              ...   
       2431424  G          T          G2055V                0.000000   
       2431518  C          A          A2086A                0.038462   
       2431527  G          C          V2089V                0.038462   
       2431569  C          A          V2103V                0.038462   
       2431613  T          C          V2118A                0.000000   

                                                 frq_BF-09_colu_2012  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                  0.006098   
       2358328  T          C          NaN                   0.006098   
       2358441  A          T          NaN                   0.000000   
       2358468  A          C          NaN                   0.000000   
       2358581  T          A          NaN                   0.000000   
...                                                              ...   
       2431424  G          T          G2055V                0.000000   
       2431518  C          A          A2086A                0.000000   
       2431527  G          C          V2089V                0.000000   
       2431569  C          A          V2103V                0.000000   
       2431613  T          C          V2118A                0.000000   

                                                 frq_BF-09_colu_2014  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                       0.0   
       

There are 9,659 rows and 18 columns in this DataFrame. Each row provides data about a SNP. The columns provide data about those SNPs, such as their effect, and the alternate allele frequency in a particular cohort of mosquitoes, e.g., the column named "frq_BF-09_colu_2012" provides the frequency in *Anopheles coluzzii* mosquitoes collected in the BF-09 (Hauts-Bassins) administrative region of Burkina Faso in 2012.

### Aside: variable naming styles

You may have noticed that in the examples above, whenever we assigned a DataFrame to a variable, the name of the variable either started with the prefix "`df_`" or ended with the suffix "`_df`". Here "df" stands for DataFrame, and is used just as a helpful reminder that the value of the variable is a pandas DataFrame. Using a prefix of suffix on the variable name like this is entirely optional and is up to your own coding style. You can name these variables whatever you like. Different instructors in this course prefer different styles, and you will see different styles being used in different modules.

### Example: reading from a CSV file

You can also create a pandas DataFrame by reading data from a file, such as a comma-separated value (CSV) file, or an Excel workbook file. Let's use the pandas [read_csv()](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html) function to read data from a CSV file containing data from a US census that colab provides as example data for you to play with.

In [9]:
df_housing = pd.read_csv("sample_data/california_housing_train.csv")
df_housing

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -114.31     34.19                15.0       5612.0          1283.0   
1        -114.47     34.40                19.0       7650.0          1901.0   
2        -114.56     33.69                17.0        720.0           174.0   
3        -114.57     33.64                14.0       1501.0           337.0   
4        -114.57     33.57                20.0       1454.0           326.0   
...          ...       ...                 ...          ...             ...   
16995    -124.26     40.58                52.0       2217.0           394.0   
16996    -124.27     40.69                36.0       2349.0           528.0   
16997    -124.30     41.84                17.0       2677.0           531.0   
16998    -124.30     41.80                19.0       2672.0           552.0   
16999    -124.35     40.54                52.0       1820.0           300.0   

       population  households  median_income  median_house_value  
0          1015.0       472.0         1.4936             66900.0  
1          1129.0       463.0         1.8200             80100.0  
2           333.0       117.0         1.6509             85700.0  
3           515.0       226.0         3.1917             73400.0  
4           624.0       262.0         1.9250             65500.0  
...           ...         ...            ...                 ...  
16995       907.0       369.0         2.3571            111400.0  
16996      1194.0       465.0         2.5179             79000.0  
16997      1244.0       456.0         3.0313            103600.0  
16998      1298.0       478.0         1.9797             85800.0  
16999       806.0       270.0         3.0147             94600.0  

[17000 rows x 9 columns]

This DataFrame has 17,000 rows and 9 columns. Each row provides data about a block of housing. The columns provide data about the houses within a block, such as the total number of bedrooms and the number of households.

Note that you can upload your own files to Colab if you want to read in and analyse some of your own data with pandas.

### Example: creating a DataFrame from scratch

You can also create a DataFrame by providing the data within your code. To demonstrate this, let's create a DataFrame with data about characters in a famous science fiction book called The Hitchhiker's Guide to the Galaxy.

In [10]:
df_hitchhikers = pd.DataFrame(data={
    "name": ["Arthur Dent", "Tricia McMillan", "Ford Prefect", "Zaphod Beeblebrox", "Marvin the Paranoid Android"],
    "taxon": ["human", "human", "betelgeusian", "betelgeusian", "robot"],
    "gender": ["male", "female", "male", "male", None],
    "number_of_heads": [1, 1, 1, 2, 1],
    "intelligence": [100, 180, 120, 30, 5_000_000]
})
df_hitchhikers

name         taxon  gender  number_of_heads  \
0                  Arthur Dent         human    male                1   
1              Tricia McMillan         human  female                1   
2                 Ford Prefect  betelgeusian    male                1   
3            Zaphod Beeblebrox  betelgeusian    male                2   
4  Marvin the Paranoid Android         robot    None                1   

   intelligence  
0           100  
1           180  
2           120  
3            30  
4       5000000

This DataFrame has 5 rows and 5 columns. Each row provides data about a fictional character in the book. The columns provide data about each character, such as their name, taxon and number of heads (some alien species can have more than one).

## Structure of a DataFrame

Hopefully this will already be obvious from the examples above, but a [pandas DataFrame](https://pandas.pydata.org/docs/user_guide/dsintro.html#dataframe) is a data structure that is organised into **rows** and **columns**. Each DataFrame also has an **index**, which is a special column (or set of columns) that is shown in bold at the left of a DataFrame, and which we'll look at in more detail later.

In [ ]:
%%html
<iframe frameborder="0" style="width:100%;height:556px;" src="https://viewer.diagrams.net/?nav=1&title=workshop-2-module-1-dataframe.drawio#Uhttps%3A%2F%2Fdrive.google.com%2Fuc%3Fid%3D1znvxOlvX2zO2dDx_BdgD7Pzztt1xS6JT%26export%3Ddownload"></iframe>

## Counting rows and columns

If a DataFrame is large, you may find it useful to know how many rows and columns it contains. The representation of a DataFrame shows at the bottom how many rows and columns are present in the DataFrame, but you can also use the built-in `len()` function if you want to access the number of rows or columns in your code.

In [11]:
num_rows = len(df_hitchhikers)
num_rows

5

In [12]:
num_cols = len(df_hitchhikers.columns)
num_cols

5

**Exercise 1 (English)**

* Count the number of rows and columns in the `df_samples` DataFrame.

**Exercice 1 (Français)**

* Compter le nombre de lignes et de colonnes du DataFrame `df_samples`.

In [13]:
df_hitchhikers.shape

(5, 5)

## Accessing columns

Often you will have a DataFrame with many columns, and you many only need to work with data from a smaller number of columns, or just a single column. Accessing one or more columns from a DataFrame can be done using the square bracket syntax after the name of a DataFrame variable. Let's see how this works.


### Accessing column names

First, let's access the names of all the columns in a DataFrame, to see what columns are available. You can do this by accessing the `columns` property of a DataFrame. For simplicity, let's also convert the result to a Python list, using the built-in `list()` function. 

In [14]:
col_names = list(df_hitchhikers.columns)
col_names

['name', 'taxon', 'gender', 'number_of_heads', 'intelligence']

This can be useful especially if you have a lot of columns in a DataFrame. 

**Exercise 2 (English)**

* Access the column names of the `df_samples` DataFrame.

**Exercice 2 (Français)**

* Afficher les noms des colonnes du DataFrame `df_samples`.

In [15]:
df_samples.columns.to_list()

['sample_id',
 'partner_sample_id',
 'contributor',
 'country',
 'location',
 'year',
 'month',
 'latitude',
 'longitude',
 'sex_call',
 'sample_set',
 'release',
 'aim_species_fraction_arab',
 'aim_species_fraction_colu',
 'aim_species_fraction_colu_no2l',
 'aim_species_gambcolu_arabiensis',
 'aim_species_gambiae_coluzzii',
 'aim_species',
 'country_iso',
 'admin1_name',
 'admin1_iso',
 'admin2_name',
 'taxon',
 'cohort_admin1_year',
 'cohort_admin1_month',
 'cohort_admin2_year',
 'cohort_admin2_month']

### Accessing a single column

To access a single column, give the column name as a string inside square brackets. Let's access the "name" column from the DataFrame of book characters, and assign the result to a new variable `character_name`.

In [16]:
character_name = df_hitchhikers["name"]

When you access a single column or row from a DataFrame, pandas returns a [Series](https://pandas.pydata.org/docs/user_guide/dsintro.html#series) object to you. Pandas uses the Series class to hold data either from a single column or a single row, i.e., a one-dimensional sequence of data values.

In [17]:
type(character_name)

pandas.core.series.Series

We can also inspect the data in the column.

In [18]:
character_name

0                    Arthur Dent
1                Tricia McMillan
2                   Ford Prefect
3              Zaphod Beeblebrox
4    Marvin the Paranoid Android
Name: name, dtype: object

In [19]:
df_hitchhikers.name

0                    Arthur Dent
1                Tricia McMillan
2                   Ford Prefect
3              Zaphod Beeblebrox
4    Marvin the Paranoid Android
Name: name, dtype: object

### Investigating data in a column

The pandas Series class has some useful functions for investigating the data in a DataFrame. 

E.g., let's access the "taxon" column and use the `value_counts()` functions to get a summary of the categorical data in the column.

In [20]:
character_taxon = df_hitchhikers["taxon"]
character_taxon.value_counts()

human           2
betelgeusian    2
robot           1
Name: taxon, dtype: int64

The pandas Series class also has a `describe()` function which can be useful for investigating numerical data. Let's use it with the `number_of_heads` column.

In [21]:
character_heads = df_hitchhikers["number_of_heads"]
character_heads.describe()

count    5.000000
mean     1.200000
std      0.447214
min      1.000000
25%      1.000000
50%      1.000000
75%      1.000000
max      2.000000
Name: number_of_heads, dtype: float64

The characters in our DataFrame have at least (min) 1 and at most (max) 2 heads. The mean number of heads is 1.2, but of course this a bit silly, because number of heads is not a continuous variable.

**Exercise 3 (English)**

* Access the "country" column from the `df_samples` DataFrame, and then use the `value_counts()` function to show the number of mosquito samples from each country.

* Access the "type" column from the `df_geneset` DataFrame, and then use the `value_counts()` function to show the number of genome annotations of different types.

* Access the "effect" column from the `snp_freqs_df` DataFrame, and then use the `value_counts()` function to show the number of SNPs with different effects.

**Exercice 3 (Français)**

* Obtenir la colonne "country" du DataFrame `df_samples` et utiliser la fonction `value_counts()` pour afficher le nombre de moustiques capturés dans chaque pays.

* Obtenir la colonne "type" du DataFrame `df_geneset` et utiliser la fonction `value_counts()` pour afficher le nombre d'annotations du génome de chaque type.

* Obtenir la colonne "effect" du DataFrame `snp_freqs_df` et utiliser la fonction `value_counts()` pour afficher le nombre de SNPs pour chaque effet possible.

In [22]:
df_samples.country.value_counts()



Cameroon                            444
Tanzania                            300
Lab Cross                           297
Burkina Faso                        296
Uganda                              290
Gambia, The                         279
Mali                                225
Guinea                              136
Guinea-Bissau                       101
Ghana                               100
Kenya                                86
Angola                               81
Cote d'Ivoire                        80
Democratic Republic of the Congo     76
Mozambique                           74
Central African Republic             73
Gabon                                69
Malawi                               41
Mayotte                              23
Equatorial Guinea                    10
Name: country, dtype: int64

In [23]:
df_geneset.head()

contig      source             type   start       end  score strand  phase  \
0     2L  VectorBase       chromosome       1  49364325    NaN    NaN    NaN   
1     2L  VectorBase             gene  157348    186936    NaN      -    NaN   
2     2L  VectorBase             mRNA  157348    181305    NaN      -    NaN   
3     2L  VectorBase  three_prime_UTR  157348    157495    NaN      -    NaN   
4     2L  VectorBase             exon  157348    157623    NaN      -    NaN   

              ID         Parent              Name  \
0             2L            NaN               NaN   
1     AGAP004677            NaN               NaN   
2  AGAP004677-RA     AGAP004677               NaN   
3            NaN  AGAP004677-RA               NaN   
4            NaN  AGAP004677-RA  AGAP004677-RB-E4   

                                         description  
0                                                NaN  
1  methylenetetrahydrofolate dehydrogenase(NAD ) ...  
2                                                NaN  
3                                                NaN  
4                                                NaN

In [24]:
df_geneset.type.value_counts()



exon                      72257
CDS                       66103
five_prime_UTR            16598
mRNA                      14979
gene                      13057
three_prime_UTR           11665
ncRNA_gene                  729
tRNA                        362
rRNA                        242
pre_miRNA                    77
snRNA                        35
pseudogene                   10
pseudogenic_transcript       10
chromosome                    8
ncRNA                         4
SRP_RNA                       3
snoRNA                        2
lnc_RNA                       2
RNase_MRP_RNA                 1
RNase_P_RNA                   1
Name: type, dtype: int64

In [25]:
snp_freqs_df.effect.value_counts()

INTRONIC                 9451
NON_SYNONYMOUS_CODING     121
SYNONYMOUS_CODING          56
STOP_GAINED                14
SPLICE_REGION               9
SPLICE_CORE                 8
Name: effect, dtype: int64

### Accessing multiple columns

To access multiple columns, provide the column names as a list of strings. Let's access the "name", "taxon" and "number_of_heads" columns from the DataFrame of book characters, and assign the resulting DataFrame to a new variable `df_hh`.

In [26]:
selected_cols = ["name", "taxon", "number_of_heads"]
df_hh = df_hitchhikers[selected_cols]
df_hh

name         taxon  number_of_heads
0                  Arthur Dent         human                1
1              Tricia McMillan         human                1
2                 Ford Prefect  betelgeusian                1
3            Zaphod Beeblebrox  betelgeusian                2
4  Marvin the Paranoid Android         robot                1

You can also put the list of columns directly inside the square brackets, which would give the same result.

In [27]:
df_hh = df_hitchhikers[["name", "taxon", "number_of_heads"]]
df_hh

name         taxon  number_of_heads
0                  Arthur Dent         human                1
1              Tricia McMillan         human                1
2                 Ford Prefect  betelgeusian                1
3            Zaphod Beeblebrox  betelgeusian                2
4  Marvin the Paranoid Android         robot                1

**Exercise 4 (English)**

* Create a new DataFrame containing only the "sample_id", "country" and "year" columns from the `df_samples` DataFrame.
* Create a new DataFrame containing only the "ID", "contig", "start" and "end" columns from the `df_geneset` DataFrame.
* Create a new DataFrame containing only the frequency columns from the `snp_freqs_df` DataFrame (hint: the names of these columns all begin with "frq_").

**Exercice 4 (Français)**

* Créer un nouveau DataFrame contenant seulement les colonnes "sample_id", "country" et "year" du DataFrame `df_samples`.

* Créer un nouveau DataFrame contenant seulement les colonnes "ID", "contig", "start" et "end" du DataFrame `df_geneset`.

* Créer un nouveau DataFrame contenant seulement les colonnes contenant les fréquences du DataFrame `snp_freqs_df` (Indice: les noms de ces colonnes commencent tous par "frq_").

In [28]:
df_samples[["sample_id","country","year"]]

sample_id    country  year
0     AR0047-C     Angola  2009
1     AR0049-C     Angola  2009
2     AR0051-C     Angola  2009
3     AR0061-C     Angola  2009
4     AR0078-C     Angola  2009
...        ...        ...   ...
3076  AD0494-C  Lab Cross    -1
3077  AD0495-C  Lab Cross    -1
3078  AD0496-C  Lab Cross    -1
3079  AD0497-C  Lab Cross    -1
3080  AD0498-C  Lab Cross    -1

[3081 rows x 3 columns]

In [29]:
df_geneset[["ID","contig","start","end"]]

ID      contig   start       end
0                  2L          2L       1  49364325
1          AGAP004677          2L  157348    186936
2       AGAP004677-RA          2L  157348    181305
3                 NaN          2L  157348    157495
4                 NaN          2L  157348    157623
...               ...         ...     ...       ...
196140            NaN  Y_unplaced   47932     48111
196141            NaN  Y_unplaced   47932     48138
196142  AGAP029375-PA  Y_unplaced   48112     48138
196143            NaN  Y_unplaced   48301     48385
196144  AGAP029375-PA  Y_unplaced   48301     48385

[196145 rows x 4 columns]

In [30]:
snp_freqs_df.loc[:, snp_freqs_df.columns.str.startswith("frq_")]

frq_BF-07_gamb_2004  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                  0.000000   
       2358328  T          C          NaN                   0.000000   
       2358441  A          T          NaN                   0.038462   
       2358468  A          C          NaN                   0.269231   
       2358581  T          A          NaN                   0.192308   
...                                                              ...   
       2431424  G          T          G2055V                0.000000   
       2431518  C          A          A2086A                0.038462   
       2431527  G          C          V2089V                0.038462   
       2431569  C          A          V2103V                0.038462   
       2431613  T          C          V2118A                0.000000   

                                                 frq_BF-09_colu_2012  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                  0.006098   
       2358328  T          C          NaN                   0.006098   
       2358441  A          T          NaN                   0.000000   
       2358468  A          C          NaN                   0.000000   
       2358581  T          A          NaN                   0.000000   
...                                                              ...   
       2431424  G          T          G2055V                0.000000   
       2431518  C          A          A2086A                0.000000   
       2431527  G          C          V2089V                0.000000   
       2431569  C          A          V2103V                0.000000   
       2431613  T          C          V2118A                0.000000   

                                                 frq_BF-09_colu_2014  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                       0.0   
       2358328  T          C          NaN                        0.0   
       2358441  A          T          NaN                        0.0   
       2358468  A          C          NaN                        0.0   
       2358581  T          A          NaN                        0.0   
...                                                              ...   
       2431424  G          T          G2055V                     0.0   
       2431518  C          A          A2086A                     0.0   
       2431527  G          C          V2089V                     0.0   
       2431569  C          A          V2103V                     0.0   
       2431613  T          C          V2118A                     0.0   

                                                 frq_BF-09_gamb_2012  \
contig position ref_allele alt_allele aa_change                        
2L     2358252  C          T          A32V                  0.000000   
       2358328  T          C          NaN                   0.000000   
       2358441  A          T          NaN                   0.000000   
       2358468  A          C          NaN                   0.000000   
       2358581  T          A          NaN                   0.000000   
...                                                              ...   
       2431424  G          T          G2055V                0.005102   
       2431518  C          A          A2086A                0.000000   
       2431527  G          C          V2089V                0.000000   
       2431569  C          A          V2103V                0.000000   
       2431613  T          C          V2118A                0.005102   

                                                 frq_BF-09_gamb_2014  
contig position ref_allele alt_allele aa_change                       
2L     2358252  C          T          A32V                       0.0  
       2358328  T          C          NaN                        0.0  
       2358441  A

## Accessing rows

Now we've seen how to access columns, let's see how to access rows. To access rows by their position within a DataFrame, we can use the `iloc` property. 



### Accessing a single row

To access a single row, we use the `iloc` property, then provide the row number as an integer inside square brackets, remembering that numbering starts from 0 in Python. For example, access the first row of our DataFrame of book characters, and assign the result to a variable `first_row`.

In [31]:
first_row = df_hitchhikers.iloc[0]

When accessing a single row, pandas returns a Series.

In [32]:
type(first_row)

pandas.core.series.Series

Inspect the data in the row.

In [33]:
first_row

name               Arthur Dent
taxon                    human
gender                    male
number_of_heads              1
intelligence               100
Name: 0, dtype: object

Access specific data values within the row.

In [34]:
first_row["name"]

'Arthur Dent'

In [35]:
first_row["number_of_heads"]

1

**Exercise 5 (English)**

* Access the 81st row in the `df_samples` DataFrame. What country did this mosquito sample come from? (Remember: numbering starts from 0 in Python.)

* Access the 73,287th row in the `df_geneset` DataFrame. What type of annotation is this? What is its Name?

**Exercice 5 (Français)**

* Obtenir la 81ème ligne du DataFrame `df_samples`. De quel pays est issu ce moustique? (Indice: les indices commencent à 0 en Python)

* Obtenir la 73 287ème ligne du DataFrame `df_geneset`. De quel type d'annotation s'agit-il? Quel est son nom?

In [36]:
df_samples.iloc[80].country

'Angola'

In [37]:
df_geneset.iloc[73286].type

'gene'

### Accessing multiple rows

We can also use `iloc` to access multiple rows,from some start position to some end position. To do this we provide two integers separated by a colon (":") which is called a slice. 

As well as remembering that numbering starts from zero, we also have to remember that slices include the start position but do not include the stop position. E.g., to access the first two rows, provide the start and stop positions 0 and 2. This returns a new DataFrame with just the selected rows within the slice.

In [38]:
first_two_rows = df_hitchhikers.iloc[0:2]
first_two_rows

name  taxon  gender  number_of_heads  intelligence
0      Arthur Dent  human    male                1           100
1  Tricia McMillan  human  female                1           180

**Exercise 6 (English)**

* Access rows starting from the 6th row up to and including the 10th row in the `df_samples` DataFrame.

**Exercice 6 (Français)**

* Obtenir toutes les lignes entre la 6ème et la 10ème (incluses) du DataFrame `df_samples`.

In [39]:
df_samples.iloc[5:10]

sample_id partner_sample_id contributor country location  year  month  \
5  AR0080-C            LUA080  Joao Pinto  Angola   Luanda  2009      4   
6  AR0084-C            LUA084  Joao Pinto  Angola   Luanda  2009      4   
7  AR0097-C            LUA097  Joao Pinto  Angola   Luanda  2009      4   
8  AR0072-C            LUA072  Joao Pinto  Angola   Luanda  2009      4   
9  AR0094-C            LUA094  Joao Pinto  Angola   Luanda  2009      4   

   latitude  longitude sex_call  ... aim_species country_iso  admin1_name  \
5    -8.884     13.302        F  ...    coluzzii         AGO       Luanda   
6    -8.884     13.302        F  ...    coluzzii         AGO       Luanda   
7    -8.884     13.302        F  ...    coluzzii         AGO       Luanda   
8    -8.884     13.302        F  ...    coluzzii         AGO       Luanda   
9    -8.884     13.302        M  ...    coluzzii         AGO       Luanda   

   admin1_iso  admin2_name     taxon cohort_admin1_year  cohort_admin1_month  \
5      AO-LUA       Luanda  coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
6      AO-LUA       Luanda  coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
7      AO-LUA       Luanda  coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
8      AO-LUA       Luanda  coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   
9      AO-LUA       Luanda  coluzzii   AO-LUA_colu_2009  AO-LUA_colu_2009_04   

        cohort_admin2_year         cohort_admin2_month  
5  AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
6  AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
7  AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
8  AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  
9  AO-LUA_Luanda_colu_2009  AO-LUA_Luanda_colu_2009_04  

[5 rows x 27 columns]

## Indexes

Every pandas DataFrame has an index, which is a special column (or set of columns) which can be used to access rows quickly and conveniently. In particular, indexes can be used to assign a label to every row, and then use that label to locate and access matching rows.

For example, when working with our DataFrame of book characters, we might want to access rows using the name of the character. To do this, we can set the "name" column as the index.

In [40]:
df_hitchhikers_indexed = df_hitchhikers.set_index("name")
df_hitchhikers_indexed

taxon  gender  number_of_heads  \
name                                                                 
Arthur Dent                         human    male                1   
Tricia McMillan                     human  female                1   
Ford Prefect                 betelgeusian    male                1   
Zaphod Beeblebrox            betelgeusian    male                2   
Marvin the Paranoid Android         robot    None                1   

                             intelligence  
name                                       
Arthur Dent                           100  
Tricia McMillan                       180  
Ford Prefect                          120  
Zaphod Beeblebrox                      30  
Marvin the Paranoid Android       5000000

The DataFrame `df_hitchhikers_indexed` has the same data as before, but notice that the "name" column has moved to the left and is shown in bold. This shows us that the "name" column is now being used as the DataFrame index.

Now that we've set a useful index, we can use the `loc` property to locate rows with a particular name. E.g., locate the row where the character's name is "Tricia McMillan".

In [41]:
trillian = df_hitchhikers_indexed.loc["Tricia McMillan"]

This indexing action locates a single row, and so pandas returns a Series. 

In [42]:
type(trillian)

pandas.core.series.Series

Inspect the row of data we located.

In [43]:
trillian

taxon               human
gender             female
number_of_heads         1
intelligence          180
Name: Tricia McMillan, dtype: object

It's also useful to know that you can always reset the index back to the default, using the `reset_index()` function. The default is to index the DataFrame with integers.

In [44]:
df_hitchhikers_reset = df_hitchhikers_indexed.reset_index()
df_hitchhikers_reset

name         taxon  gender  number_of_heads  \
0                  Arthur Dent         human    male                1   
1              Tricia McMillan         human  female                1   
2                 Ford Prefect  betelgeusian    male                1   
3            Zaphod Beeblebrox  betelgeusian    male                2   
4  Marvin the Paranoid Android         robot    None                1   

   intelligence  
0           100  
1           180  
2           120  
3            30  
4       5000000

**Exercise 7 (English)**

* Set the "sample_id" column as the index on the `df_samples` DataFrame, then locate the row with sample identifier "BK0055-C". Where was this mosquito collected?
* Set the "ID" column as the index on the `df_geneset` DataFrame, then locate the row with identifier "AGAP000818". What is the name of this gene?

**Exercice 7 (Français)**

* Faire de la colonne "sample_id" l'index du DataFrame `df_samples` et ensuite trouver la ligne avec comme identificant d'échantillon "BK0055-C". D'où est issu ce moustique?

* Faire de la colonne "ID" l'index du DataFrame `df_geneset` et ensuite trouver la ligne avec comme identifiant "AGAP000818". Quel est le nom de ce gène?


In [45]:
df_samples.set_index("sample_id").loc["BK0055-C"]

partner_sample_id                                      RCA_62
contributor                            Alessandra della Torre
country                              Central African Republic
location                                               Bangui
year                                                     1994
month                                                       1
latitude                                                4.367
longitude                                              18.583
sex_call                                                    F
sample_set                                         AG1000G-CF
release                                                   3.0
aim_species_fraction_arab                            0.000958
aim_species_fraction_colu                            0.017167
aim_species_fraction_colu_no2l                       0.019769
aim_species_gambcolu_arabiensis                      gambcolu
aim_species_gambiae_coluzzii                          gambiae
aim_spec

In [46]:
df_geneset.set_index("ID").loc["AGAP000818"]

contig                                                        X
source                                               VectorBase
type                                                       gene
start                                                  15240572
end                                                    15242864
score                                                       NaN
strand                                                        -
phase                                                       NaN
Parent                                                      NaN
Name                                                     CYP9K1
description    cytochrome P450 [Source:VB Community Annotation]
Name: AGAP000818, dtype: object

## Queries

DataFrames also provide a `query()` function, which can be used to select rows matching some condition or combination of conditions. The `query()` function accepts a string containing a query expression.

Let's query the book characters DataFrame to select only rows where the value of the "taxon" column is "human". To match a value we use the "==" operator.


In [47]:
df_human_hitchhikers = df_hitchhikers.query("taxon == 'human'")
df_human_hitchhikers

name  taxon  gender  number_of_heads  intelligence
0      Arthur Dent  human    male                1           100
1  Tricia McMillan  human  female                1           180

Let's do the opposite, selecting only rows where the value of the "taxon" column is not "human", using the "!=" operator.

In [48]:
df_non_human_hitchhikers = df_hitchhikers.query("taxon != 'human'")
df_non_human_hitchhikers

name         taxon gender  number_of_heads  \
2                 Ford Prefect  betelgeusian   male                1   
3            Zaphod Beeblebrox  betelgeusian   male                2   
4  Marvin the Paranoid Android         robot   None                1   

   intelligence  
2           120  
3            30  
4       5000000

Let's query a numerical field using the ">" greater than operator.

In [49]:
df_multi_headed_hitchhikers = df_hitchhikers.query("number_of_heads > 1")
df_multi_headed_hitchhikers

name         taxon gender  number_of_heads  intelligence
3  Zaphod Beeblebrox  betelgeusian   male                2            30

A query expression can combine conditions on multiple columns using "and" and "or". E.g., find rows where the value of the "taxon" column is "human" and the value of the "number_of_heads" column is 1.

In [50]:
df_non_human_one_headed_hitchhikers = df_hitchhikers.query(
    "taxon != 'human' and number_of_heads == 1"
)
df_non_human_one_headed_hitchhikers

name         taxon gender  number_of_heads  \
2                 Ford Prefect  betelgeusian   male                1   
4  Marvin the Paranoid Android         robot   None                1   

   intelligence  
2           120  
4       5000000

**Exercise 8 (English)**

* Query the `df_samples` DataFrame to select mosquitoes collected in Burkina Faso.
* Query the `df_samples` DataFrame to select mosquitoes collected in Burkina Faso in 2014.
* Query the `df_samples` DataFrame to select Anopheles coluzzii mosquitoes collected in Burkina Faso in 2014 (hint: `taxon == 'coluzzii'`).
* Query the `snp_freqs_df` DataFrame to select SNPs which change the protein sequence (hint: `effect == 'NON_SYNONYMOUS_CODING'`).
* Query the `snp_freqs_df` DataFrame to select SNPs which change the protein sequence and which occur above 5% frequency in at least one cohort (hint: `max_af > 0.05`).

**Exercice 8 (Français)**

* Sélectionner parmi les moustiques du DataFrame `df_samples` ceux qui ont été capturés au Burkina Faso.

* Sélectionner parmi les moustiques du DataFrame `df_samples` ceux qui ont été capturés au Burkina Faso en 2014.

* Sélectionner parmi les moustiques du DataFrame `df_samples` les Anophèles coluzzii qui ont été capturés au Burkina Faso en 2014. (Indice: `taxon == 'coluzzii'`).

* Sélectionner parmi les SNPs du DataFrame `snp_freqs_df` ceux qui modifient la séquence d'une protéine (Indice: `effect == 'NON_SYNONYMOUS_CODING'`).

* Sélectionner parmi les SNPs du DataFrame `snp_freqs_df` ceux qui modifient la séquence d'une protéine et apparaissent avec une fréquence supérieure à 5% dans au moins une cohorte (Indice: `max_af > 0.05`).


In [51]:
list(df_samples.columns)

['sample_id',
 'partner_sample_id',
 'contributor',
 'country',
 'location',
 'year',
 'month',
 'latitude',
 'longitude',
 'sex_call',
 'sample_set',
 'release',
 'aim_species_fraction_arab',
 'aim_species_fraction_colu',
 'aim_species_fraction_colu_no2l',
 'aim_species_gambcolu_arabiensis',
 'aim_species_gambiae_coluzzii',
 'aim_species',
 'country_iso',
 'admin1_name',
 'admin1_iso',
 'admin2_name',
 'taxon',
 'cohort_admin1_year',
 'cohort_admin1_month',
 'cohort_admin2_year',
 'cohort_admin2_month']

In [52]:
df_samples.country.unique()

array(['Angola', 'Burkina Faso', 'Democratic Republic of the Congo',
       'Central African Republic', "Cote d'Ivoire", 'Cameroon', 'Mayotte',
       'Gabon', 'Ghana', 'Gambia, The', 'Guinea', 'Mali',
       'Equatorial Guinea', 'Guinea-Bissau', 'Kenya', 'Malawi',
       'Mozambique', 'Tanzania', 'Uganda', 'Lab Cross'], dtype=object)

In [53]:
# Query the df_samples DataFrame to select mosquitoes collected in Burkina Faso.
# Query the df_samples DataFrame to select mosquitoes collected in Burkina Faso in 2014.
# Query the df_samples DataFrame to select Anopheles coluzzii mosquitoes collected in Burkina Faso in 2014 (hint: taxon == 'coluzzii').

burkina_faso = df_samples.query("country == 'Burkina Faso'")
burkina_faso_2014 = burkina_faso.query("year == 2014")
df_samples.query("country == 'Burkina Faso' and year == 2014 and taxon == 'coluzzii'")

sample_id partner_sample_id  contributor       country        location  \
262  AB0326-C            BF18-1  Austin Burt  Burkina Faso    Bana Village   
263  AB0327-C            BF18-3  Austin Burt  Burkina Faso    Bana Village   
264  AB0328-C            BF18-4  Austin Burt  Burkina Faso    Bana Village   
265  AB0329-C            BF18-5  Austin Burt  Burkina Faso    Bana Village   
266  AB0330-C            BF18-6  Austin Burt  Burkina Faso    Bana Village   
267  AB0331-C            BF18-9  Austin Burt  Burkina Faso    Bana Village   
268  AB0332-C           BF18-10  Austin Burt  Burkina Faso    Bana Village   
270  AB0334-C           BF18-12  Austin Burt  Burkina Faso    Bana Village   
271  AB0337-C           BF22-15  Austin Burt  Burkina Faso    Bana Village   
272  AB0338-C           BF18-13  Austin Burt  Burkina Faso    Bana Village   
273  AB0339-C           BF18-14  Austin Burt  Burkina Faso    Bana Village   
274  AB0340-C           BF18-15  Austin Burt  Burkina Faso    Bana Village   
275  AB0341-C           BF18-16  Austin Burt  Burkina Faso    Bana Village   
276  AB0348-C           BF22-50  Austin Burt  Burkina Faso    Bana Village   
277  AB0351-C            BF23-6  Austin Burt  Burkina Faso    Bana Village   
278  AB0352-C           BF18-46  Austin Burt  Burkina Faso    Bana Village   
279  AB0354-C           BF18-21  Austin Burt  Burkina Faso    Bana Village   
280  AB0355-C           BF18-22  Austin Burt  Burkina Faso    Bana Village   
281  AB0356-C           BF18-23  Austin Burt  Burkina Faso    Bana Village   
282  AB0357-C           BF18-24  Austin Burt  Burkina Faso    Bana Village   
283  AB0358-C           BF18-25  Austin Burt  Burkina Faso    Bana Village   
284  AB0359-C           BF18-26  Austin Burt  Burkina Faso    Bana Village   
285  AB0360-C           BF18-28  Austin Burt  Burkina Faso    Bana Village   
286  AB0361-C           BF18-29  Austin Burt  Burkina Faso    Bana Village   
287  AB0362-C           BF18-31  Austin Burt  Burkina Faso    Bana Village   
288  AB0363-C           BF18-32  Austin Burt  Burkina Faso    Bana Village   
289  AB0364-C           BF18-33  Austin Burt  Burkina Faso    Bana Village   
290  AB0365-C           BF18-34  Austin Burt  Burkina Faso    Bana Village   
291  AB0366-C           BF18-35  Austin Burt  Burkina Faso    Bana Village   
292  AB0367-C           BF18-36  Austin Burt  Burkina Faso    Bana Village   
293  AB0368-C           BF18-37  Austin Burt  Burkina Faso    Bana Village   
294  AB0369-C           BF18-38  Austin Burt  Burkina Faso    Bana Village   
295  AB0370-C           BF18-42  Austin Burt  Burkina Faso    Bana Village   
296  AB0371-C           BF18-44  Austin Burt  Burkina Faso    Bana Village   
297  AB0372-C           BF18-45  Austin Burt  Burkina Faso    Bana Village   
307  AB0396-C            BF14-8  Austin Burt  Burkina Faso    Bana Village   
308  AB0397-C            BF14-9  Austin Burt  Burkina Faso    Bana Village   
309  AB0398-C           BF14-10  Austin Burt  Burkina Faso    Bana Village   
310  AB0399-C           BF14-11  Austin Burt  Burkina Faso    Bana Village   
311  AB0400-C           BF14-12  Austin Burt  Burkina Faso    Bana Village   
312  AB0403-C           BF14-16  Austin Burt  Burkina Faso    Bana Village   
313  AB0404-C           BF14-18  Austin Burt  Burkina Faso    Bana Village   
314  AB0409-C           BF14-21  Austin Burt  Burkina Faso    Bana Village   
315  AB0410-C           BF20-54  Austin Burt  Burkina Faso    Bana Village   
316  AB0411-C           BF20-56  Austin Burt  Burkina Faso    Bana Village   
317  AB0413-C           BF20-59  Austin Burt  Burkina Faso    Bana Village   
342  AB0503-C            BF13-1  Austin Burt  Burkina Faso  Souroukoudinga   
344  AB0505-C           BF13-13  Austin Burt  Burkina Faso  Souroukoudinga   
345  AB0508-C           BF13-38  Austin Burt  Burkina Faso  Souroukoudinga   
346  AB0509-C           BF24-55  Austin Burt  Burkina Faso  Souroukoudinga   
347  AB0510-C           BF

In [54]:
# Query the snp_freqs_df DataFrame to select SNPs which change the protein sequence (hint: effect == 'NON_SYNONYMOUS_CODING').
# Query the snp_freqs_df DataFrame to select SNPs which change the protein sequence and which occur above 5% frequency in at least one cohort (hint: max_af > 0.05).

snp_freqs_df.query("effect == 'NON_SYNONYMOUS_CODING'")
snp_freqs_df.query("effect == 'NON_SYNONYMOUS_CODING' and max_af > 0.05")

pass_gamb_colu_arab  \
contig position ref_allele alt_allele aa_change                        
2L     2391228  G          C          V402L                     True   
                           T          V402L                     True   
       2416980  C          T          T791M                     True   
       2422652  A          T          L995F                     True   
       2429617  T          C          I1527T                    True   
       2429745  A          T          N1570Y                    True   
       2429897  A          G          E1597G                    True   
       2429915  A          C          K1603T                    True   
       2430424  G          T          A1746S                   False   
       2430863  T          C          I1868T                    True   
       2430880  C          T          P1874S                    True   
       2430881  C          T          P1874L                    True   
       2431061  C          T          A1934V                    True   

                                                 pass_gamb_colu  pass_arab  \
contig position ref_allele alt_allele aa_change                              
2L     2391228  G          C          V402L                True       True   
                           T          V402L                True       True   
       2416980  C          T          T791M                True       True   
       2422652  A          T          L995F                True       True   
       2429617  T          C          I1527T               True       True   
       2429745  A          T          N1570Y               True       True   
       2429897  A          G          E1597G               True       True   
       2429915  A          C          K1603T               True       True   
       2430424  G          T          A1746S               True      False   
       2430863  T          C          I1868T               True       True   
       2430880  C          T          P1874S               True       True   
       2430881  C          T          P1874L               True       True   
       2431061  C          T          A1934V               True       True   

                                                 frq_BF-07_gamb_2004  \
contig position ref_allele alt_allele aa_change                        
2L     2391228  G          C          V402L                 0.000000   
                           T          V402L                 0.000000   
       2416980  C          T          T791M                 0.000000   
       2422652  A          T          L995F                 0.076923   
       2429617  T          C          I1527T                0.000000   
       2429745  A          T          N1570Y                0.038462   
       2429897  A          G          E1597G                0.000000   
       2429915  A          C          K1603T                0.000000   
       2430424  G          T          A1746S                0.000000   
       2430863  T          C          I1868T                0.000000   
       2430880  C          T          P1874S                0.000000   
       2430881  C          T          P1874L                0.000000   
       2431061  C          T          A1934V                0.000000   

                                                 frq_BF-09_colu_2012  \
contig position ref_allele alt_allele aa_change                        
2L     2391228  G          C          V402L                 0.067073   
                           T          V402L                 0.054878   
       2416980  C          T          T791M                 0.018293   
       2422652  A          T          L995F                 0.865854   
       2429617  T          C          I1527T                0.121951   
       2429745  A          T          N1570Y                0.250000   
       2429897  A          G          E1597G                0.000000   
       2429915  A          C          K1603T                0.054878

## Input/output

We saw at the beginning of this module how to read in data from a file. Let's look finally at how to write data out to a file.

To write a DataFrame out to a CSV file, use the [to_csv()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_csv.html) function. E.g., let's write out the book characters DataFrame to a CSV file.


In [55]:
df_hitchhikers.to_csv("hitchhikers.csv")

You can download this file to your local computer by opening the file browser tab to the left in colab.

There are several other file formats you can write to. E.g., you could write to an Excel file.

In [56]:
df_hitchhikers.to_excel("hitchhikers.xlsx")

Any of the DataFrames we created in this module could be written out in the same way. E.g., we could write a DataFrame that we obtained by querying the book characters for only humans.

In [57]:
df_human_hitchhikers.to_excel("hh_human.xlsx")

**Exercise 9 (English)**

* Write the `df_samples` DataFrame out to an Excel file.
* Query the `df_samples` DataFrame to select only mosquitoes from Burkina Faso, then write the resulting DataFrame out to a CSV file.

**Exercice 9 (Français)**

* Enregistrer le DataFrame `df_samples` sous forme de fichier Excel.
* Sélectionner parmi les moustiques du DataFrame `df_samples` ceux provenant du Burkina Faso et enregistrer le DataFrame ainsi créé sous forme de fichier CSV.

In [58]:
df_samples.to_excel("df_samples.xlsx")
df_samples.query("country == 'Burkina Faso'").to_excel("burkina_faso.xlsx")

## Topics not covered here

Please note that this module has only covered a few of the many features that pandas supports. Here are a few of the features that we have not covered, but which you might like to learn more about through other online tutorials.

* Grouping and aggregation
* Pivot tables
* Joining and merging
* Concatenating
* Advanced indexing (multi-level indexes; Boolean indexing; ...)
* Modifying data
* Apply
* Missing data
* Time series
* Plotting
* Text
* ...


## Further reading

Here are a few links to online resources providing further information about pandas for new users.

* [pandas getting started](https://pandas.pydata.org/pandas-docs/stable/getting_started/index.html)
* [pandas user guide](https://pandas.pydata.org/docs/user_guide/index.html) (in particular see [10 minutes to pandas](https://pandas.pydata.org/docs/user_guide/10min.html))
* [pandas community tutorials](https://pandas.pydata.org/pandas-docs/stable/getting_started/tutorials.html)